libnavigation-util/src/main/java/com/mapbox/navigation/utils/timer/MapboxTimer.kt

```diff
@@ -0,0 +1,31 @@
+package com.mapbox.navigation.utils.timer
+
+import com.mapbox.navigation.utils.thread.ThreadController
+import kotlinx.coroutines.delay
+import kotlinx.coroutines.isActive
+import kotlinx.coroutines.launch
+
+/**
+ * Schedules a delay of [restartAfter] milliseconds and then restarts.
+ *
+ * @param restartAfter Time delay until the timer should restart.
+ * @param delayLambda lambda function that is to be executed after [restartAfter] milliseconds.
+ */
+class MapboxTimer(private val restartAfter: Long, private val delayLambda: () -> Unit) {
+    private val mainControllerJobScope = ThreadController.getMainScopeAndRootJob()
+
+    fun start() {
+        mainControllerJobScope.scope.launch {
+            while (isActive) {
+                delay(restartAfter)
+                delayLambda()
+            }
+        }
+    }
+
+    fun stop() {
+        mainControllerJobScope.job.children.forEach { job ->
```

Comment:
NIT

```suggestion
        mainControllerJobScope.job.children.forEach { it.cancel() }
```

Reply:
I like using `names` instead of `it`. It gives more clarity and readability for anyone new to this coding paradigm.

Reply:
For a one liner, I think it's 🆗 / cleaner. Also, for consistency (we're using that across other different classes).

Reply:
Actually I've just realized that we're not exposing a way to cancel `mainControllerJobScope`. As part of `stop` we're cancelling all its children but what about itself? If using `MapboxTimer` is it safe to not cancel the parent (`mainControllerJobScope`)? Should we expose a kill API? cc @abhishek1508 @olegzil

Reply:
just `mainControllerJobScope.scope.cancel()` should be called.
Because it cancels this scope, including its job and all its children
@Guardiola31337

Reply:
> just `mainControllerJobScope.scope.cancel()` should be called.
Because it cancels this scope, including its job and all its children

Yeah, I know, there are some situations in which you may want to start / stop the timer as part of the same lifecycle, if we do `mainControllerJobScope.scope.cancel()` (as it's currently implemented) next call to `start` will crash. We could potentially workaround this changing the implementation and creating / destroying `mainControllerJobScope` as part of `start` / `stop` OR provide a third `kill` API as mentioned above. What do you think @abhishek1508 @olegzil @korshaknn?

Reply:
@Guardiola31337  in `start` method we can recreate scope if it was cancelled 
if (!mainControllerJobScope.scope.isActive) {
    mainControllerJobScope = ThreadController.getMainScopeAndRootJob()
}

Reply:
Yeah, as I mentioned those are the options available. Looking forward to hearing what you think @olegzil @abhishek1508 @korshaknn

Reply:
@Guardiola31337 could you cut a ticket so that this doesn't get forgotten?